### PR TITLE
 Integrate Pydantic Basemodels for gmso.abc (Release 0.3.0.)

### PR DESCRIPTION
**Pull request by umesh-timalsina**
Checklist (Overall)
This checklist is subject to change. As this is a major overhaul, It might take a while to make it active.
- [x] **Module `gmso.abc`**
  - [x] Base Class with Configuration
  - [x] Abstract `Site`
  - [x] Abstract `Connection`
  - [x] Abstract `Potential`
  - [x] AutoDocs for pydantic base classes (Inspired and Customized from `QCElemental`)
-  [x] **Core Changes**
   - [x] Rename `gmso.Site` to `gmso.Atom` and Inherit from `Site`
   - [x] Inherit `gmso.Bond`, `gmso.Angle`, `gmso.Dihedral`, `gmso.Improper` from `Connection`.
   - [x] `Potential` Classes inherit from `gmso.abc.Potential`
   - [x] `Element`, `PotentialTemplates` changes
- [x] **Formats and External Changes**  
   - [x] No changes should be necessary As per my understanding. 
- [x] **Benchmarks** (Pydantic vs current Master)

### CODE REVIEW

**Comment by umesh-timalsina** [comment]
The tests are failing beause of the use of id based hashing. See, https://github.com/samuelcolvin/pydantic/issues/1837 and a potential workaround in https://github.com/samuelcolvin/pydantic/issues/265. Use this workaround in the `GMSOBase` worked for now.

**Comment by umesh-timalsina** [comment]
There are couple of changes remaining for this PR:

1. Refactor `gmso.core.Potential` to `gmso.core.ParametricPotential`  which inherit from `gmso.abc.abstract_potential.AbstractPotential`
2. Refactor inherited classes using `gmso.core.ParametricPotential`. (`AtomType`, `BondType`, `AngleType`, `DihedralType`,  `ImproperType`).

**Comment by umesh-timalsina** [comment]
I am going to delegate changing Topology and handling serialization aspects to a different PR as 40 file changes are I think sufficient for a single PR. This PR is officially ready for review and revisions.

File: gmso/tests/test_atom.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        lithium_type = AtomType(mass=6.941, charge=1)
+        atom = Atom(name='Lithium')
+        atom.atom_type = lithium_type
+
+        assert atom.charge == 1*u.elementary_charge
+        assert atom.mass == 6.941*u.gram/u.mol
+
+    @pytest.mark.parametrize('position', [[0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0], [[0.0, 0.0], [0.0, 0.0]],
+        ['a', 'b', 'c'], ['a', 1, 1]])
```
Comment (justinGilmer) [inline gmso/tests/test_atom.py:61]: can we add one test where some of the values in the array are `NaN` and others are 0.0?

File: gmso/abc/abstract_connection.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    @property
+    def name(self):
+        return self.__dict__.get('name_')
+
+    @root_validator(pre=True)
+    def inject_name(cls, values):
+        connection_members = values.get('connection_members')
+        if len(set(connection_members)) != len(connection_members):
+            raise GMSOError(f'A {cls.__name__} between same {type(connection_members[0]).__name__}s is not allowed')
```
Comment (justinGilmer) [inline gmso/abc/abstract_connection.py:36]: Maybe we can mention that the user is trying to form a bond with itself

File: gmso/abc/abstract_potential.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    """
+
+    name_: str = Field(
+        '',
+        description='The name of the potential. Defaults to class name'
+    )
+
+    potential_expression_: _PotentialExpression = Field(
+        _PotentialExpression(expression='a*x+b', independent_variables={'x'}),
+        description='The potential expression for the potential'
```
Comment (justinGilmer) [inline gmso/abc/abstract_potential.py:28]:
```suggestion
        description='The mathematical expression of the potential'
```

File: gmso/abc/abstract_site.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    Site is the object that represents any general interaction site in a molecular simulation.
+    Sites have been designed to be as general as possible, making no assumptions about representing atoms or beads, or
+    having mass or charge. That is, a Site can represent an atom in an atomistic system,
+    a bead in a coarse-grained system, and much more. 
+
+    Notes
+    -----
+    The label attribute for a site takes its meaning when used with some sort of container (like topology)
+    such that a label for a site can then be used to group sites together. The rules for defining a site label
+    and their meaning is left upto the container where the sites will reside.
```
Comment (justinGilmer) [inline gmso/abc/abstract_site.py:32]:
```suggestion
    and their meaning is the responsibility to the container where the sites will reside. 
```

File: gmso/abc/abstract_site.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    )
+
+    label_: str = Field(
+        '',
+        description='Label to be assigned for the site'
+    )
+
+    position_: PositionType = Field(
+        u.unyt_array([np.nan]*3, u.nm),
+        description='The Cartesian coordinates of the position of the site'
```
Comment (justinGilmer) [inline gmso/abc/abstract_site.py:47]:
```suggestion
        description='The 3D Cartesian coordinates of the position of the site'
```

File: gmso/abc/abstract_site.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    def is_valid_position(cls, position):
+        """Validator for attribute position"""
+        if position is None:
+            return u.unyt_array([np.nan] * 3, u.nm)
+
+        if not isinstance(position, u.unyt_array):
+            try:
+                position *= u.nm
+            except InvalidUnitOperation as e:
+                raise GMSOError('Converting object of type {0} failed with following error: {1}'
```
Comment (justinGilmer) [inline gmso/abc/abstract_site.py:75]: We use an `fstring` in the ValueError below, could we use fstrings here, or is there a reason not to?

File: gmso/abc/abstract_site.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+            warnings.warn('Positions are assumed to be in nm')
+
+        input_unit = position.units
+
+        position = np.asarray(position, dtype=float, order='C')
+        try:
+            position = np.reshape(position, newshape=(3,), order='C')
+        except ValueError:
+            raise ValueError(f'Position of shape {position.shape} is not valid. '
+                             'Accepted values: (a.) 3-tuple, (b.) list of length 3 '
```
Comment (justinGilmer) [inline gmso/abc/abstract_site.py:86]: would `list-like of length 3` work for `a` and `b`?

File: gmso/abc/abstract_site.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    such that a label for a site can then be used to group sites together. The rules for defining a site label
+    and their meaning is left upto the container where the sites will reside. 
+    """
+
+    name_: str = Field(
+        '',
+        description='Name of the site, defaults to class name',
+    )
+
+    label_: str = Field(
```
Comment (justinGilmer) [inline gmso/abc/abstract_site.py:40]: It might be useful to have multiple labels associated with a Site

For example, I might want to label an Atom with multiple labels based on my search criteria. 

Maybe it should be a set of strings?

File: gmso/core/angle.py [hunk 9]
```diff
@@ -1,55 +1,56 @@
...
-
-
-def _validate_angletype(contype):
-    """Ensure connection_type is a AngleType """
-    if contype is None:
-        warnings.warn("Non-parametrized Angle detected")
-    elif not isinstance(contype, AngleType):
-        raise GMSOError("Supplied non-AngleType {}".format(contype))
-    return contype
+    connection_members_: Tuple[Atom, Atom, Atom] = Field(
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/angle.py:24]: Yeah, it should be possible:
image

File: gmso/core/angle.py [hunk 10]
```diff
@@ -1,55 +1,56 @@
...
+    @property
+    def angle_type(self):
+        return self.__dict__.get('angle_type_')
+
+    @property
+    def connection_type(self):
+        return self.__dict__.get('angle_type_')
+
+    def __setattr__(self, key, value):
+        if key == 'connection_type':
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/angle.py:43]: Yeah. `connection.connection_type` is used in other avenues in the code base.

File: gmso/core/atom.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    @validator('charge_')
+    def is_valid_charge(cls, charge):
+        if charge is None:
+            return None
+        if not isinstance(charge, u.unyt_array):
+            warnings.warn("Charges are assumed to be elementary charge")
+            charge *= u.elementary_charge
+
+        elif charge.units.dimensions != u.elementary_charge.units.dimensions:
+            warnings.warn("Charges are assumed to be elementary charge")
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/atom.py:91]: Agreed. I will change it.

File: gmso/core/parametric_potential.py [hunk 12]
```diff
@@ -195,5 +151,14 @@ def from_template(cls, potential_template, parameters, topology=None):
...
-                   template=False)
+                   topology=topology)
+
+    class Config:
+        fields = {
+            'topology_': 'topology',
+            'set_ref_': 'set_ref'
+        }
+        alias_to_fields = {
+            'topology': 'topology_'
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/parametric_potential.py:162]: The set_ref variable is an internal constant and is exposed as readonly property. It may be provided during class initialization, however, cannot be changed using `__setattr__`

File: gmso/tests/test_atom.py (hunk 1 above)
Comment (umesh-timalsina) [reply gmso/tests/test_atom.py:61]: We can do that but, I don't think we should do it in this PR, may be we should open this as an issue and add it there?

File: gmso/abc/abstract_site.py (hunk 6 above)
Comment (umesh-timalsina) [reply gmso/abc/abstract_site.py:75]: No, not really, this was copied as is from the code in master,

File: gmso/abc/abstract_site.py (hunk 8 above)
Comment (umesh-timalsina) [reply gmso/abc/abstract_site.py:40]: Agreed. Should we do this separately??

File: gmso/core/parametric_potential.py [hunk 13]
```diff
@@ -141,29 +114,12 @@ def set_expression(self, expression=None, parameters=None, independent_variables
         If only a subset of the parameters are supplied, they are updated
         while the non-passed parameters default to the existing values
        """
-        self._potential_expression.set(
+        self.potential_expression_.set(
```
Comment (uppittu11) [inline gmso/core/parametric_potential.py:117]: Is the potential expression validated upon assignment? Or will that be part of another PR?

File: gmso/core/atom.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    gmso.abc.AbstractSite
+        An Abstract Base class for implementing site objects in GMSO. The class Atom bases from
+        the gmso.abc.abstract site class
+    """
+    charge_: Optional[Union[u.unyt_quantity, float]] = Field(
+        None,
+        description='Charge of the atom',
+    )
+
+    mass_: Optional[Union[u.unyt_quantity, float]] = Field(
```
Comment (uppittu11) [inline gmso/core/atom.py:37]: unyt_quantities/unyt_arrays are not json serializable right now. That's probably not in the scope of this PR, but it is holding up our ability to serialize Atoms (and other classes) at the moment.

Comment (uppittu11) [reply gmso/core/atom.py:37]: NVM you mentioned this in a comment earlier

File: gmso/core/parametric_potential.py (hunk 13 above)
Comment (umesh-timalsina) [reply gmso/core/parametric_potential.py:117]: It will be validated. However, it might be nice to check if its actually parametric i.e. it has parameters set correctly.

File: gmso/core/subtopology.py [hunk 15]
```diff
@@ -79,7 +79,7 @@ def add_site(self, site):
 
         Parameters
         ----------
-        site : gmso.Site
+        site : gmso.Atom
```
Comment (daico007) [inline gmso/core/subtopology.py:82]: we might want this to support general `site` subclasses

File: gmso/external/convert_mbuild.py [hunk 16]
```diff
@@ -85,7 +85,7 @@ def from_mbuild(compound, box=None, search_method=element_by_symbol):
             for particle in child.particles():
                 pos = particle.xyz[0] * u.nanometer
                 ele = search_method(particle.name)
-                site = Site(name=particle.name, position=pos, element=ele)
+                site = Atom(name=particle.name, position=pos, element=ele)
```
Comment (daico007) [inline gmso/external/convert_mbuild.py:88]: This can be another PR, but we can do the checking when converting from mbuild and see if we want to create a GMSO bead or atom (now that Compound has that element attribute)

File: gmso/external/convert_parmed.py [hunk 17]
```diff
@@ -65,14 +65,14 @@ def from_parmed(structure, refer_type=True):
         subtops.append(gmso.SubTopology(name=subtop_name, parent=top))
         for atom in residue.atoms:
             if refer_type and isinstance(atom.atom_type, pmd.AtomType):
-                site = gmso.Site(
+                site = gmso.Atom(
```
Comment (daico007) [inline gmso/external/convert_parmed.py:68]: same as mbuild conversion

File: gmso/core/element.py [hunk 18]
```diff
@@ -2,47 +2,51 @@
...
 import unyt as u
 
+from gmso.abc.gmso_base import GMSOBase
 from gmso.exceptions import GMSOError
 from gmso.utils.misc import unyt_to_hashable
 
-class Element(namedtuple('Element', 'atomic_number, name, symbol, mass')):
-    """Chemical element object
+
+class Element(GMSOBase):
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/element.py:13]: Yeah see #435.

File: gmso/abc/abstract_potential.py [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    @property
+    def expression(self):
+        return self.potential_expression_.expression
+
+    @property
+    def potential_expression(self):
+        return self.__dict__.get('potential_expression_')
+
+    def set_expression(self):
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/abc/abstract_potential.py:73]: This method is supposed to be overridden in the downstream classes which inherit from AbstractPotential. This is good find, there should be a decoration on this method.

```py
@abstract_method
def set_expression(self):
    raise NotImplementedError
```

File: gmso/core/angle.py [hunk 20]
```diff
@@ -1,55 +1,56 @@
...
+
+from gmso.core.atom import Atom
+from gmso.abc.abstract_connection import Connection
 from gmso.core.angle_type import AngleType
-from gmso.exceptions import GMSOError
 
 
 class Angle(Connection):
-    """A 3-partner connection between sites.
+    __base_doc__ = """A 3-partner connection between Atoms.
```
[the remark this answers is not in the document]
Comment (uppittu11) [reply gmso/core/angle.py:11]: I don't remember if we have any other documentation on the subclasses of `Site`, but [here](https://hackmd.io/05ZiOPs3T8uS0AA0pqMJTw) it looks like we decided that `VirtualSite` and `Point` shouldn't be able to participate in connections. IDK how we want to handle this. Should there be an intermediate subclass between `Site` and `Atom`/`Bead` called `ConnectableSite` or something? 

This change is probably better suited for another PR tho.

[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/angle.py:11]: Technically, I think this `Angle` can be extended with something like this:
```py
connection_members = Union[
     Tuple[Atom, Atom, Atom],
     Tuple[JustAnotherConnetableSite, JustAnotherConnectableSite, JustAnotherConnetableSite]
]
```
For this to work, we can and have defined where the angle Actually exists and all the combinations can therefore be validatied.

[the remark this answers is not in the document]
Comment (umesh-timalsina) [reply gmso/core/angle.py:11]: However, I agree with @uppittu11. Probably will need another PR to add this feature.

File: gmso/core/atom.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    __base_doc__ = """An atom represents a single element association in a topology.
+    
+    Atoms are the representation of an element within `gmso` that describes any general 
+    atom in a molecular simulation. Atoms also contain information that are unique to 
+    elements vs other types of interaction sites in molecular simulations. 
+    For example, charge, mass, and periodic table information.
+
+    Notes
+    -----
+    Atoms have all the attributes inherited from the base Site class
```
Comment (justinGilmer) [inline gmso/core/atom.py:24]: Should add that "Order of precedence when attaining charge: atom.charge -> atom.atom_type.charge -> None"

Comment (justinGilmer) [reply gmso/core/atom.py:24]: Same for mass

File: gmso/core/atom.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+    @validator('charge_')
+    def is_valid_charge(cls, charge):
+        if charge is None:
+            return None
+        if not isinstance(charge, u.unyt_array):
+            warnings.warn(UNIT_WARNING_STRING.format('Charges', 'elementary charge'))
+            charge *= u.elementary_charge
+
+        elif charge.units.dimensions != u.elementary_charge.units.dimensions:
+            warnings.warn(UNIT_WARNING_STRING.format('Charges', 'elementary charge'))
```
Comment (justinGilmer) [inline gmso/core/atom.py:96]: This warning would not be the exact same case. The first case is when no unyt there, the second case is when the unyt "unit" does not simplify to a "charge" unit dimension.

Comment (justinGilmer) [reply gmso/core/atom.py:96]: Second case here would be an error. Units are in a unit that cannot be coerced into charge. (length != charge)

File: gmso/core/atom.py [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+        return charge
+
+    @validator('element_')
+    def is_valid_element(cls, element):
+        if element is not None:
+            assert isinstance(element, Element), 'Attribute element must be of type gmso.core.Element'
+        return element
+
+    @validator('mass_')
+    def is_valid_mass(cls, mass):
```
Comment (justinGilmer) [inline gmso/core/atom.py:108]: same comments as above for charge

File: gmso/core/atom.py [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+            charge *= u.elementary_charge
+
+        elif charge.units.dimensions != u.elementary_charge.units.dimensions:
+            warnings.warn(UNIT_WARNING_STRING.format('Charges', 'elementary charge'))
+            charge = charge.value * u.elementary_charge
+
+        return charge
+
+    @validator('element_')
+    def is_valid_element(cls, element):
```
Comment (justinGilmer) [inline gmso/core/atom.py:102]: this is defined above with the FieldType Annotation for the element itself. Should not be needed

File: gmso/core/bond.py [hunk 25]
```diff
@@ -1,54 +1,56 @@
...
-
-
-def _validate_bondtype(contype):
-    """Ensure connection_type is a BondType """
-    if contype is None:
-        warnings.warn("Non-parametrized Bond detected")
-    elif not isinstance(contype, BondType):
-        raise GMSOError("Supplied non-BondType {}".format(contype))
-    return contype
+    connection_members_: Tuple[Atom, Atom] = Field(
```
[the remark this answers is not in the document]
Comment (justinGilmer) [reply gmso/core/bond.py:23]: Answered with: https://github.com/mosdef-hub/gmso/pull/431/files#r481382382

File: gmso/core/bond.py [hunk 26]
```diff
@@ -1,54 +1,56 @@
...
+        description='BondType of this bond.'
+    )
+
+    @property
+    def bond_type(self):
+        return self.__dict__.get('bond_type_')
+
+    @property
+    def connection_type(self):
+        # ToDo: Deprecate this?
```
Comment (justinGilmer) [inline gmso/core/bond.py:39]: We can deprecate this a bit later.

File: gmso/core/angle.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (justinGilmer) [reply gmso/core/angle.py:24]: answered by: https://github.com/mosdef-hub/gmso/pull/431/files#r481382382

File: gmso/core/dihedral.py [hunk 27]
```diff
@@ -17,43 +19,43 @@ class Dihedral(Connection):
...
-
-
-def _validate_dihedraltype(contype):
-    """Ensure connection_type is a DihedralType """
-    if contype is None:
-        warnings.warn("Non-parametrized Dihedral detected")
-    elif not isinstance(contype, DihedralType):
-        raise GMSOError("Supplied non-DihedralType {}".format(contype))
-    return contype
+    connection_members_: Tuple[Atom, Atom, Atom, Atom] = Field(
```
[the remark this answers is not in the document]
Comment (justinGilmer) [reply gmso/core/dihedral.py:28]: https://github.com/mosdef-hub/gmso/pull/431/files#r481382382

File: gmso/core/improper.py [hunk 28]
```diff
@@ -21,44 +23,44 @@ class Improper(Connection):
...
-    name : str, optional, default=Improper
-        Name of the improper.
-
     Notes
     -----
     Inherits some methods from Connection:
         __eq__, __repr__, _validate methods
     Additional _validate methods are presented
     """
+    connection_members_: Tuple[Atom, Atom, Atom, Atom] = Field(
```
[the remark this answers is not in the document]
Comment (justinGilmer) [reply gmso/core/improper.py:32]: https://github.com/mosdef-hub/gmso/pull/431/files#r481382382

**Review by justinGilmer — CHANGES_REQUESTED** [review]
Part 1 of N group reviews @umesh-timalsina @uppittu11 @justinGilmer @rmatsum836 @daico007

File: gmso/core/atom_type.py [hunk 29]
```diff
@@ -76,90 +76,53 @@ def __init__(self,
         if overrides is None:
             overrides = set()
 
+        kwargs = {
```
Comment (justinGilmer) [inline gmso/core/atom_type.py:79]: instead of using `kwargs` below, we could access them from the dict.

File: gmso/core/atom_type.py [hunk 30]
```diff
@@ -176,50 +139,51 @@ def __hash__(self):
...
-        raise ValueError("Passed overrides of non-string to overrides")
-    return val
+    @validator('mass_', pre=True)
+    def validate_mass(cls, mass):
+        """Check to see that a mass is a unyt array of the right dimension"""
+        if not isinstance(mass, u.unyt_array):
+            warnings.warn(UNIT_WARNING_STRING.format('Masses', 'g/mol'))
+            mass *= u.gram / u.mol
+        elif mass.units.dimensions != (u.gram / u.mol).units.dimensions:
+            warnings.warn(UNIT_WARNING_STRING.format('Masses', 'g/mol'))
```
Comment (justinGilmer) [inline gmso/core/atom_type.py:149]: from previous review, we should error in case of dimensionality mismatch

File: gmso/core/atom_type.py [hunk 31]
```diff
@@ -19,44 +20,43 @@ class AtomType(Potential):
...
+    mass_: Optional[u.unyt_array] = Field(
+        0.0 * u.gram / u.mol,
+        description='The mass of the atom type'
+    )
+
+    charge_: Optional[u.unyt_array] = Field(
+        0.0 * u.elementary_charge,
+        description='The charge of the atom type'
+    )
+
```
Comment (justinGilmer) [inline gmso/core/atom_type.py:34]: Should this contain an `element` attribute? There is an `element` field in the FF XMLs, but would this just be the name, or a string that can later be coerced into an element, etc.

File: gmso/core/angle_type.py [hunk 32]
```diff
@@ -18,31 +18,17 @@ class AngleType(Potential):
...
     Notes
     ----
     Inherits many functions from gmso.Potential:
         __eq__, _validate functions
-
     """
 
+    member_types_: Optional[Tuple[str, str, str]] = Field(
+        None,
+        description='List-like of gmso.AtomType.name defining the members of this angle type'
```
Comment (justinGilmer) [inline gmso/core/angle_type.py:29]: or atomclass, #442

File: gmso/core/bond_type.py [hunk 33]
```diff
@@ -18,35 +18,24 @@ class BondType(Potential):
...
     Notes
     ----
     Inherits many functions from gmso.Potential:
         __eq__, _validate functions
-
     """
 
+    member_types_: Optional[Tuple[str, str]] = Field(
+        None,
+        description='List-like of of gmso.AtomType.name defining the members of this bond type'
```
Comment (justinGilmer) [inline gmso/core/bond_type.py:29]: or atomclass #442

File: gmso/core/dihedral_type.py [hunk 34]
```diff
@@ -22,38 +24,24 @@ class DihedralType(Potential):
...
     Notes
     ----
     Inherits many functions from gmso.Potential:
         __eq__, _validate functions
-
     """
 
+    member_types_: Optional[Tuple[str, str, str, str]] = Field(
+        None,
+        description='List-like of of gmso.AtomType.name defining the members of this dihedral type'
```
Comment (justinGilmer) [inline gmso/core/dihedral_type.py:35]: or atomclass #442

File: gmso/core/improper_type.py [hunk 35]
```diff
@@ -26,39 +28,24 @@ class ImproperType(Potential):
...
     Notes
     ----
     Inherits many functions from gmso.Potential:
         __eq__, _validate functions
-
     """
 
+    member_types_: Optional[Tuple[str, str, str, str]] = Field(
+        None,
+        description='List-like of of gmso.AtomType.name defining the members of this improper type'
```
Comment (justinGilmer) [inline gmso/core/improper_type.py:39]: or atomclass #442

File: gmso/core/parametric_potential.py [hunk 36]
```diff
@@ -15,36 +19,26 @@ class Potential(object):
...
-            default={'a': 1.0*u.dimensionless, 'b': 1.0*u.dimensionless}
-        The parameters of the potential and their values, as unyt quantities.
-        The keys are names of the variables included in `expression` and values
-        are the numerical values of these parameters recorded as instances of
-        `unyt.unyt_quantity`, which combine both value and unit information.
-    independent_variables : str or sympy.Symbol or list or set thereof
-        The independent variables in the expression of the potential.
-    topology: gmso.core.Topology, the topology of which this potential is a part of, default=None
-    set_ref: (str), the string name of the bookkeeping set in topology class.
+    topology_: Optional[Any] = Field(
```
Comment (justinGilmer) [inline gmso/core/parametric_potential.py:24]: `Optional[Any]` should be gmso.topology

**Review by justinGilmer — CHANGES_REQUESTED** [review]
Notes from group review @umesh-timalsina @daico007 @uppittu11 @justinGilmer

File: gmso/external/convert_mbuild.py (hunk 16 above)
Comment (umesh-timalsina) [reply gmso/external/convert_mbuild.py:88]: Got it. This might be better to use after that version of mbuild is released

File: gmso/core/parametric_potential.py (hunk 36 above)
Comment (umesh-timalsina) [reply gmso/core/parametric_potential.py:25]: Fixed with validator. Delegating this for later issue/pr.

File: gmso/utils/misc.py [hunk 37]
```diff
@@ -22,3 +23,21 @@ def _unyt_to_hashable_single(val):
         return tuple(val.in_base().value)
     else:
         return None
+
+
+def ensure_valid_dimensions(quantity_1: u.unyt_quantity,
+                            quantity_2: u.unyt_quantity) -> None:
+    """Ensure that the quantities provided as arguments have valid dimensions
```
Comment (umesh-timalsina) [inline gmso/utils/misc.py:30]: We need a better description for this method. Something like this suggested by @justinGilmer .

```suggestion
    This utility will verify that the two unyt quantities provided simplify to the same dimensionality. This ensures that each quantity can be converted to the desired unit system.
```

**Review by justinGilmer — APPROVED** [review]

**Review by daico007 — APPROVED** [review]
